esphome/components/display_menu_base/display_menu_base.cpp

```diff
@@ -0,0 +1,554 @@
+    this->displayed_item_ = this->displayed_item_->get_parent();
+    this->top_index_ = this->selection_stack_.front().first;
+    this->cursor_index_ = this->selection_stack_.front().second;
+    this->selection_stack_.pop_front();
+    this->displayed_item_->on_enter();
+    chg = true;
+  }
+
+  return chg;
+}
+
+void DisplayMenuComponent::finish_editing_() {
+  switch (this->get_selected_item_()->get_type()) {
+    case MENU_ITEM_SELECT:
+    case MENU_ITEM_NUMBER:
+    case MENU_ITEM_SWITCH:
+    case MENU_ITEM_CUSTOM:
+      this->get_selected_item_()->on_leave();
+      break;
+    default:
+      break;
+  }
+
+  this->editing_ = false;
+}
+
+void DisplayMenuComponent::draw_menu() {
+  if (this->active_) {
+    for (size_t i = 0; i < this->rows_ && this->top_index_ + i < this->displayed_item_->items_size(); ++i) {
+      this->draw_item(this->displayed_item_->get_item(this->top_index_ + i), i,
+                      this->top_index_ + i == this->cursor_index_);
+    }
+  }
+}
+
+bool MenuItem::select_next() {
+  bool val;
+
+  switch (this->item_type_) {
+#ifdef USE_SELECT
+    case MENU_ITEM_SELECT:
+      val = this->next_option_();
+      break;
+#endif
+#ifdef USE_NUMBER
+    case MENU_ITEM_NUMBER:
+      val = this->inc_number_();
+      break;
+#endif
+#ifdef USE_SWITCH
+    case MENU_ITEM_SWITCH:
+      val = this->toggle_switch_();
+      break;
+#endif
+    case MENU_ITEM_COMMAND:
+      this->on_value_();
+      val = true;
+      break;
+    case MENU_ITEM_CUSTOM:
+      this->on_next_();
+      this->on_value_();
+      val = true;
+      break;
+    default:
+      val = false;
+      break;
+  }
+
+  return val;
+}
+
+bool MenuItem::select_prev() {
+  bool val;
+
+  switch (this->item_type_) {
+#ifdef USE_SELECT
+    case MENU_ITEM_SELECT:
+      val = this->prev_option_();
+      break;
+#endif
+#ifdef USE_NUMBER
+    case MENU_ITEM_NUMBER:
+      val = this->dec_number_();
+      break;
+#endif
+#ifdef USE_SWITCH
+    case MENU_ITEM_SWITCH:
+      val = this->toggle_switch_();
+      break;
+#endif
+    case MENU_ITEM_COMMAND:
+      this->on_value_();
+      val = true;
+      break;
+    case MENU_ITEM_CUSTOM:
+      this->on_prev_();
+      this->on_value_();
+      val = true;
+      break;
+    default:
+      val = false;
+      break;
+  }
+
+  return val;
+}
+
+void MenuItem::on_enter() { this->on_enter_callbacks_.call(); }
+
+void MenuItem::on_leave() { this->on_leave_callbacks_.call(); }
+
+#ifdef USE_SELECT
+bool MenuItem::next_option_() {
+  bool chg = false;
+
+  if (this->select_var_ != nullptr) {
+    this->select_var_->make_call().select_next(true).perform();
+    chg = true;
+  }
+
+  return chg;
+}
+
+bool MenuItem::prev_option_() {
+  bool chg = false;
+
+  if (this->select_var_ != nullptr) {
+    this->select_var_->make_call().select_previous(true).perform();
+    chg = true;
+  }
+
+  return chg;
+}
+#endif  // USE_SELECT
+
+#ifdef USE_NUMBER
+bool MenuItem::inc_number_() {
+  bool chg = false;
```

Comment:
I just chose a random one, but please use full names for all variables, it only serves as confusion when trying to read the code.
```suggestion
  bool changed = false;
```